### PR TITLE
swap args to H5Sselect_hyperslab for better efficiency

### PR DESCRIPTION
**Pull request by khou2020**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

Update HDF5 module to call H5Sselect_hyperslab in the intended way.
According to the HDF5 manual (https://support.hdfgroup.org/HDF5/doc/RM/H5S/H5Sselect_hyperslab.htm), the fifth argument (count) means the number of hyper-slabs to select while the sixth argument (block) is the size of each hyper-slab.
In this case, qmcpack is writing a single hyper-slab in a dataset, so the intended way to call H5Sselect_hyperslab is H5Sselect_hyperslab(space_id, H5S_SELECT_SET, offsets, NULL, one, blocks)

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Ubuntu

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -1,6 +1,44 @@
 #!/bin/bash
-echo Configuring and building froom inside the build directory.
-echo Check the results of the CMake configuration to ensure that the preferred
-echo compilers and libraries have been selected. See README and documentation 
-echo for guidance.
-cd build; cmake ..; make; cd ..
+
+#BUILD_MODULES=config/load_olcf_summit_modules.sh
```
Comment (ye-luo) [inline build.sh:3]: Do not touch this file.

File: src/QMCHamiltonians/ObservableHelper.cpp [hunk 2]
```diff
@@ -147,7 +147,10 @@ void ObservableHelper::write(const value_type* first_v, const value_type* /*firs
   if (rank)
   {
     H5Sset_extent_simple(space1_id, rank, &curdims[0], &maxdims[0]);
-    H5Sselect_hyperslab(space1_id, H5S_SELECT_SET, &offsets[0], NULL, &mydims[0], NULL);
+    hsize_t *one = (hsize_t*)malloc(sizeof(hsize_t) * rank);
+    for (int i = 0; i < rank; i++)  one[i] = 1;
```
Comment (ye-luo) [inline src/QMCHamiltonians/ObservableHelper.cpp:151]: const std::vector<hsize_t> one(rank, 1);

**Review by ye-luo — CHANGES_REQUESTED** [review]

File: src/QMCHamiltonians/ObservableHelper.cpp (hunk 2 above)
Comment (khou2020) [reply src/QMCHamiltonians/ObservableHelper.cpp]: fixed

File: build.sh (hunk 1 above)
Comment (khou2020) [reply build.sh]: fixed.

File: src/QMCHamiltonians/ObservableHelper.cpp [hunk 3]
```diff
@@ -147,7 +147,8 @@ void ObservableHelper::write(const value_type* first_v, const value_type* /*firs
   if (rank)
   {
     H5Sset_extent_simple(space1_id, rank, &curdims[0], &maxdims[0]);
-    H5Sselect_hyperslab(space1_id, H5S_SELECT_SET, &offsets[0], NULL, &mydims[0], NULL);
+    const std::vector<hsize_t> one(rank, 1);
```
Comment (PDoakORNL) [inline src/QMCHamiltonians/ObservableHelper.cpp:150]: Document this, I just see a vector called one that does seem to have some size and is all ones. It's basically a magic number.  It's sort of explained in your PR but it should be documented here.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
I'm not following how this isn't a breaking change for anything that reads what is being written here.  Perhaps the HDF5 API is really complex enough that these two calls result in data that can be read for the hd5 file with the same call?

File: src/QMCHamiltonians/ObservableHelper.cpp (hunk 3 above)
Comment (khou2020) [reply src/QMCHamiltonians/ObservableHelper.cpp:150]: > I'm not following how this isn't a breaking change for anything that reads what is being written here. Perhaps the HDF5 API is really complex enough that these two calls result in data that can be read for the hd5 file with the same call?

The net effect is the same. In the origin version, it is telling HDF5 that you are selecting mydims hyper-slabs along each dimensions with each hyper-slab of unit size 1x1x... . In this PR, it is selecting a single hyper-slab of size mydims.

In summary:
H5Sselect_hyperslab (*,*,*,*, count, NULL);
is equivalent (in terms final result) to
H5Sselect_hyperslab (*,*,*,*, one, count);
but they have different high-level meaning, and the later one is closer to what qmcpack intended to do.

Comment (khou2020) [reply src/QMCHamiltonians/ObservableHelper.cpp:150]: > Document this, I just see a vector called one that does seem to have some size and is all ones. It's basically a magic number. It's sort of explained in your PR but it should be documented here.

Should I put the explanation as comments before every H5Sselect_hyperslab?
I think it is intuitive given the spec of H5Sselect_hyperslab.
The vector of one means we want to write one hyper-slab (block) along each of the dimensions and each hyper-slab is of size count.
The original version write count hyper-slabs along each dimension and each hyper-slab is of size one.
They lead to the same result but literally means different objective.

Comment (prckent) [reply src/QMCHamiltonians/ObservableHelper.cpp:150]: Thanks for these welcome improvements. Perhaps just comment the first usage in each file. Can also put this PR URL.

**Comment by prckent** [comment]
Test this please

**Review by prckent — COMMENTED** [review]
LGTM but I'll wait for Peter or Ye to approve. CI is running.

**Review by PDoakORNL — APPROVED** [review]

**Comment by wkliao** [comment]
Any plan to incorporate this PR?
It improves performance quite a bit.

**Comment by ye-luo** [comment]
> Any plan to incorporate this PR? It improves performance quite a bit.

I was asking @khou2020 performance data at scale and got no rely. If you have alternative data to show, that is also OK.

**Review by prckent — APPROVED** [review]
While I would prefer to see some performance data at scale, I also recognize that it can be difficult to come by.

@ye-luo Since this would be easy to unwind in case any problems show up (say, some weirdness on GPFS), how about we merge?

**Comment by ye-luo** [comment]
Since @wkliao mentioned "It improves performance quite a bit." He should have some numbers to tell us. The code seems OK but our CI doesn't check parallel HDF5. I will need to run restart tests manually once.

**Comment by wkliao** [comment]
@khou2020 and I are developing a new HDF5 [log-based VOL](https://github.com/DataLib-ECP/vol-log-based)
and using QMCPACK to evaluate it. The performance difference between
with this PR and without is significant. One read case shows 65 seconds
vs. 1 second.

Just to clarify, this PR is more of correcting QMCPACK's mis-use of 
`H5Sselect_hyperslab`. If you checked the [HDF5 reference manual](https://support.hdfgroup.org/HDF5/doc/RM/H5S/H5Sselect_hyperslab.htm),
it describes the difference of arguments count and block. Also see the example
provide there which helps understand it. When argument stride is NULL, the
argument count used in QMCPACK should be all 1s and argument block
should be the argument count.

`H5Sselect_hyperslab` is not only used in parallel I/O.
It affects serial I/O as well.

**Review by ye-luo — APPROVED** [review]
Restart tests passed.

**Comment by ye-luo** [comment]
Test this please

**Comment by prckent** [comment]
Test this please

**Comment by ye-luo** [comment]
The CI failure is unrelated. The Clang14 offload runs do have sporadic failures. I will trigger failed test again.
